karma/karma-puppeteer-launcher/index.js

```diff
@@ -60,7 +63,59 @@ function puppeteerBrowser(baseBrowserDecorator, config) {
   });
 }
 
-async function exposeFunctions(page) {
+async function exposeFunctions(page, config) {
```

Comment:
I just learned about `@percy/puppeteer` ([repo](https://github.com/percy/percy-puppeteer) | [docs](https://docs.percy.io/docs/puppeteer) | [example project](https://github.com/percy/example-percy-puppeteer)).

This would avoid temporarily adding snapshots to `.test_artifacts`.

Thoughts?

Reply:
I think this could work great and, at the very least, a good validation for our approach. But couple of nuances:

1. `percySnapshot` takes a `Page` type, but we need a `Frame`. They are mostly interchangeable, but this is risky: even if it works now, it could break at any point later. I filed this request at them: https://github.com/percy/percy-puppeteer/issues/174
2. It works by inserting a script on the page that does the actual snapshots. We will have a lot of frames with our tests, so I'm not sure how much this may slow down the tests - we will have to try.
3. We will need to wrap our test runner within `percy exec`.
4. There are some questions w.r.t. what to do when tests fails. I think ideally we'd want to not run Percy at all to avoid "these snapshots are no longer available" for the tests that were skipped as a collateral damage. Weirdly I don't see `percy exec` options to control this. But maybe it works like that by default anyway?
5. `percySnapshot(..., { widths: [768, 992, 1200] })` sounds pretty interesting. Not 100% this works for us, but maybe it does.

Some of this would just need to be tried, and this is a promising approach. What do you think? I really think the first problem is really the only problematic one.